### PR TITLE
More "packed" behavior when reducing GameServerSet replicas.

### PR DESCRIPTION
**Pull request by castaneai**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, press enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind breaking
> /kind bug
> /kind cleanup
> /kind documentation

/kind feature

> /kind hotfix

**What this PR does / Why we need it**:
"Packed" scheduling almost always works correctly, but sometimes the GameServer to be deleted is chosen from multiple nodes at random when multiple nodes have the same number of replicas.

For more "packed" behavior, sort the GameServers to be deleted so that one node is emptied first.

**Which issue(s) this PR fixes**:

**Special notes for your reviewer**:

### CODE REVIEW

File: pkg/gameserversets/gameserversets_test.go [hunk 1]
```diff
@@ -35,13 +35,19 @@ func TestSortGameServersByLeastFullNodes(t *testing.T) {
...
 
 	list := []*agonesv1.GameServer{
 		{ObjectMeta: metav1.ObjectMeta{Name: "g1"}, Status: agonesv1.GameServerStatus{NodeName: "n2", State: agonesv1.GameServerStateReady}},
 		{ObjectMeta: metav1.ObjectMeta{Name: "g2"}, Status: agonesv1.GameServerStatus{NodeName: "", State: agonesv1.GameServerStateReady}},
 		{ObjectMeta: metav1.ObjectMeta{Name: "g3"}, Status: agonesv1.GameServerStatus{NodeName: "n1", State: agonesv1.GameServerStateReady}},
 		{ObjectMeta: metav1.ObjectMeta{Name: "g4"}, Status: agonesv1.GameServerStatus{NodeName: "n2", State: agonesv1.GameServerStateCreating}},
+		{ObjectMeta: metav1.ObjectMeta{Name: "g5"}, Status: agonesv1.GameServerStatus{NodeName: "n3", State: agonesv1.GameServerStateReady}},
+		{ObjectMeta: metav1.ObjectMeta{Name: "g6"}, Status: agonesv1.GameServerStatus{NodeName: "n4", State: agonesv1.GameServerStateReady}},
+		{ObjectMeta: metav1.ObjectMeta{Name: "g7"}, Status: agonesv1.GameServerStatus{NodeName: "n3", State: agonesv1.GameServerStateReady}},
+		{ObjectMeta: metav1.ObjectMeta{Name: "g8"}, Status: agonesv1.GameServerStatus{NodeName: "n4", State: agonesv1.GameServerStateReady}},
```
Comment (gongmax) [inline pkg/gameserversets/gameserversets_test.go:50]: Is the relative order of g5 and g7, g6 and g8 guaranteed, given `sort.Slice` is not stable?

File: pkg/gameserversets/gameserversets.go [hunk 2]
```diff
@@ -52,8 +52,12 @@ func sortGameServersByLeastFullNodes(list []*agonesv1.GameServer, count map[stri
 				return false
 			}
 		}
-
-		return (ac.Allocated + ac.Ready) < (bc.Allocated + bc.Ready)
+		// if both Nodes have the same count, one node is emptied first (packed scheduling behavior)
+		acTotal, bcTotal := ac.Allocated+ac.Ready, bc.Allocated+bc.Ready
+		if acTotal == bcTotal {
```
Comment (markmandel) [inline pkg/gameserversets/gameserversets.go:57]: Nit:

```suggestion
		if (ac.Allocated+ac.Ready) == (bc.Allocated+bc.Ready) {
```
Seems like the intermediary variables aren't really necessary? WDYT?

**Review by markmandel — COMMENTED** [review]
Good catch!

File: pkg/gameserversets/gameserversets_test.go (hunk 1 above)
Comment (markmandel) [reply pkg/gameserversets/gameserversets_test.go:50]: Run the test a few hundred times and see? 😁

Comment (castaneai) [reply pkg/gameserversets/gameserversets_test.go:50]: Surely sort.Slice is not stable!
I have run this several hundred times and the results are stable, however there are no guarantees so I will change the assertion. Thanks!

File: pkg/gameserversets/gameserversets_test.go [hunk 3]
```diff
@@ -51,6 +57,10 @@ func TestSortGameServersByLeastFullNodes(t *testing.T) {
 	assert.Equal(t, "g3", result[1].ObjectMeta.Name)
 	assert.Equal(t, "g4", result[2].ObjectMeta.Name)
 	assert.Equal(t, "g1", result[3].ObjectMeta.Name)
+	// gs on the same node are adjacent (g5 and g7, g6 and g8).
+	// The order among the two is not stable, since sort.Slice is.
+	assert.Equal(t, result[4].Status.NodeName, result[5].Status.NodeName)
```
[the remark this answers is not in the document]
Comment (castaneai) [reply pkg/gameserversets/gameserversets_test.go:62]: Looks good!

**Review by gongmax — APPROVED** [review]

**Pull request merged**